hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/management/ReadMetricsOperation.java

```diff
@@ -40,31 +38,39 @@ public ReadMetricsOperation(long offset, int collectionIntervalSeconds) {
     }
 
     @Override
-    public Object getResponse() {
-        return resultSet;
+    public void beforeRun() {
+        JetMetricsService service = getService();
+        service.getLiveOperationRegistry().register(this);
     }
 
     @Override
-    public String getServiceName() {
-        return JetMetricsService.SERVICE_NAME;
+    public void run() {
+        JetMetricsService service = getService();
+        CompletableFuture<RingbufferSlice<Entry<Long, byte[]>>> future = service.readMetrics(offset);
+        future.whenComplete((slice, error) -> doSendResponse(error != null ? error : slice));
     }
 
     @Override
-    public WaitNotifyKey getWaitKey() {
-        JetMetricsService service = getService();
-        return service.waitNotifyKey();
+    public boolean returnsResponse() {
+        return false;
     }
 
     @Override
-    public boolean shouldWait() {
-        JetMetricsService service = getService();
-        this.resultSet = service.readMetrics(offset);
-        return resultSet.elements().isEmpty();
+    public Object getResponse() {
+        throw new UnsupportedOperationException();
     }
 
     @Override
-    public void onWaitExpire() {
-        sendResponse(new OperationTimeoutException("No metrics were retrieved for " + getWaitTimeout() + "ms"));
+    public String getServiceName() {
+        return JetMetricsService.SERVICE_NAME;
     }
 
+    final void doSendResponse(Object value) {
```

Comment:
seems duplicate of AsyncJobOperation , but I see that also extends AbstractJobOperation so probably it's not possible to refactor it cleanly.